tensorflow_addons/seq2seq/tests/decoder_test.py

```diff
@@ -15,169 +15,127 @@
-                }
-            )
-
-            # Decoder only runs out to max_out; ensure values are identical
-            # to dynamic_rnn, which also zeros out outputs and passes along
-            # state.
-            self.assertAllClose(
-                eval_result["final_decoder_outputs"].rnn_output,
-                eval_result["final_rnn_outputs"][:, 0:max_out, :],
-            )
-            if use_sequence_length:
-                self.assertAllClose(
-                    eval_result["final_decoder_state"], eval_result["final_rnn_state"]
-                )
-
-    def testDynamicDecodeRNNWithTrainingHelperMatchesDynamicRNNWithSeqLen(self):
-        self._testDynamicDecodeRNNWithTrainingHelperMatchesDynamicRNN(
-            use_sequence_length=True
+def _test_decode_rnn(time_major, maximum_iterations):
+
+    sequence_length = [3, 4, 3, 1, 0]
+    batch_size = 5
+    max_time = 8
+    input_depth = 7
+    cell_depth = 10
+    max_out = max(sequence_length)
+
+    if time_major:
+        inputs = np.random.randn(max_time, batch_size, input_depth).astype(np.float32)
+    else:
+        inputs = np.random.randn(batch_size, max_time, input_depth).astype(np.float32)
+    input_t = tf.constant(inputs)
+    cell = tf.keras.layers.LSTMCell(cell_depth)
+    sampler = sampler_py.TrainingSampler(time_major=time_major)
+    my_decoder = basic_decoder.BasicDecoder(
+        cell=cell,
+        sampler=sampler,
+        output_time_major=time_major,
+        maximum_iterations=maximum_iterations,
+    )
+
+    initial_state = cell.get_initial_state(batch_size=batch_size, dtype=tf.float32)
+    (final_outputs, unused_final_state, final_sequence_length,) = my_decoder(
+        input_t, initial_state=initial_state, sequence_length=sequence_length
+    )
+
+    def _t(shape):
+        if time_major:
+            return (shape[1], shape[0]) + shape[2:]
+        return shape
+
+    # Mostly a smoke test
+    time_steps = max_out
+    expected_length = sequence_length
+    if maximum_iterations is not None:
+        time_steps = min(max_out, maximum_iterations)
+        expected_length = [min(x, maximum_iterations) for x in expected_length]
+    if tf.executing_eagerly() and maximum_iterations != 0:
+        assert (
+            _t((batch_size, time_steps, cell_depth)) == final_outputs.rnn_output.shape
         )
-
-    def testDynamicDecodeRNNWithTrainingHelperMatchesDynamicRNNNoSeqLen(self):
-        self._testDynamicDecodeRNNWithTrainingHelperMatchesDynamicRNN(
-            use_sequence_length=False
+        assert _t((batch_size, time_steps)) == final_outputs.sample_id.shape
+    np.testing.assert_array_equal(expected_length, final_sequence_length)
+
+
+@pytest.mark.usefixtures("maybe_run_functions_eagerly")
+@pytest.mark.parametrize("maximum_iterations", [None, 0, 1])
+@pytest.mark.parametrize("time_major", [True, False])
+def test_dynamic_decode_rnn(time_major, maximum_iterations):
+    _test_decode_rnn(time_major, maximum_iterations)
```

Comment:
This is not needed to have two functions here since you grouped everything, it's readable with just one.